{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"1BqDHnCEqzaxpHwYAVyPzOcZYlYGHTyDk","timestamp":1758033034912}]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["#**DATA COLLECTION**"],"metadata":{"id":"sH9NAOdX-5Ns"}},{"cell_type":"markdown","source":["Import necessary packages and modules\n"],"metadata":{"id":"LVieL8YfuulJ"}},{"cell_type":"code","source":["import requests\n","from requests.adapters import HTTPAdapter, Retry\n","import json\n","import re"],"metadata":{"id":"h3Zraokvuvx0","executionInfo":{"status":"ok","timestamp":1758286653006,"user_tz":-120,"elapsed":5,"user":{"displayName":"Aurora Mazzoni","userId":"09614232835557211324"}}},"execution_count":13,"outputs":[]},{"cell_type":"markdown","source":["Define global variables to handle the connection with the UniProt website. We use the Session object to allow retries in case of temporary service unavaiability. We set 5 as tha max number of retries."],"metadata":{"id":"dY5uwrpmvsUH"}},{"cell_type":"code","source":["retries = Retry(total=5, backoff_factor=0.25, status_forcelist=[500, 502, 503, 504])\n","session = requests.Session()\n","session.mount(\"https://\", HTTPAdapter(max_retries=retries))"],"metadata":{"id":"5PvTl4XFvWK8","executionInfo":{"status":"ok","timestamp":1758286653034,"user_tz":-120,"elapsed":4,"user":{"displayName":"Aurora Mazzoni","userId":"09614232835557211324"}}},"execution_count":14,"outputs":[]},{"cell_type":"markdown","source":["Define functions to handle API calls and pagination (useful when number of entries exceed the current limit)"],"metadata":{"id":"CMKsMgTPv54B"}},{"cell_type":"code","source":["def get_next_link(headers):\n","    if \"Link\" in headers:\n","        # The regular expression is used to extract the next link for pagination\n","        re_next_link = re.compile(r'<(.+)>; rel=\"next\"')\n","        match = re_next_link.match(headers[\"Link\"])\n","        if match:\n","            return match.group(1)\n","\n","# This function actually retrieve the next data batch in the search.\n","# The function act as an iterator, yielding the next result batch at every call\n","# The function terminates after the last batch has been returned.\n","# In this case, the next link will be None\n","def get_batch(batch_url):\n","    while batch_url:\n","        # Run the API call\n","        response = session.get(batch_url)\n","        # Will raise an error if an error status code is obtained\n","        response.raise_for_status()\n","        # Get the total number of entries in the search\n","        total = response.headers.get(\"x-total-results\", 0)\n","        # Yield the response and the total number of entries\n","        yield response, total\n","        # Get the link to the API call for the next data batch\n","        batch_url = get_next_link(response.headers)"],"metadata":{"id":"r8griYGrvUR5","executionInfo":{"status":"ok","timestamp":1758286653043,"user_tz":-120,"elapsed":4,"user":{"displayName":"Aurora Mazzoni","userId":"09614232835557211324"}}},"execution_count":15,"outputs":[]},{"cell_type":"markdown","source":["#**POSITIVE DATASET**"],"metadata":{"id":"CSnyJuhr-sVs"}},{"cell_type":"markdown","source":["#### Data Input"],"metadata":{"id":"o3BGwC-1Al1C"}},{"cell_type":"code","source":["# We define a basic URL for the search.\n","# The URL has been generated from the UniProtKB website, using the Advanced search function.\n","url_positive = \"https://rest.uniprot.org/uniprotkb/search?format=json&query=%28taxonomy_id%3A2759%29+AND+%28length%3A%5B40+TO+*%5D%29+AND+%28fragment%3Afalse%29+AND+%28reviewed%3Atrue%29+AND+%28ft_signal_exp%3A*%29&size=500\""],"metadata":{"id":"vYt64i-4AroN","executionInfo":{"status":"ok","timestamp":1758286653055,"user_tz":-120,"elapsed":4,"user":{"displayName":"Aurora Mazzoni","userId":"09614232835557211324"}}},"execution_count":16,"outputs":[]},{"cell_type":"markdown","source":["#### Function for entry filtering\n","##### It takes as input the an entry and returns it only if all the requests are satisfied"],"metadata":{"id":"V9FJUxI9ALbJ"}},{"cell_type":"code","source":["def filter_entry_positive(entry):\n","    # We iterate over the features of the entry\n","    if \"features\" in entry:\n","        for feature in entry[\"features\"]:\n","            # We only consider features of type Signal Peptides\n","            if feature[\"type\"] == \"Signal\":\n","                # Check if the lentgh is >14\n","                if type(feature[\"location\"][\"end\"][\"value\"]) == int:\n","                  if not feature.get(\"description\") :\n","                    if feature[\"location\"][\"end\"][\"value\"] >= 14 :\n","                      return True\n","    return False"],"metadata":{"id":"WHOaExvyUjdq","executionInfo":{"status":"ok","timestamp":1758286653087,"user_tz":-120,"elapsed":26,"user":{"displayName":"Aurora Mazzoni","userId":"09614232835557211324"}}},"execution_count":17,"outputs":[]},{"cell_type":"markdown","source":["#### Function extracting specific fields for each entry.\n","##### It takes as input the an entry and returns the extracted fields and the protein sequence."],"metadata":{"id":"ga3j4RWWCbXf"}},{"cell_type":"code","source":["def extract_fields_positive(entry):\n","    # We extract Uniprot accession, organism name, eukaryotic kingdom, protein length, position of signal peptide cleavage site\n","    # We iterate over the features of the entry\n","    protein_id = entry[\"primaryAccession\"] #Uniprot accession\n","    organism_name = entry[\"organism\"][\"scientificName\"] #organism name\n","    kingdom = entry[\"organism\"][\"lineage\"][1] #eukaryotic kingdom\n","    protein_length = entry[\"sequence\"][\"length\"] #protein length\n","    fasta_line = \">\"\n","    fasta_line += entry[\"primaryAccession\"]\n","    fasta_line += \"\\n\"\n","    fasta_line += entry[\"sequence\"][\"value\"] #protein sequence\n","    for feature in entry[\"features\"]:\n","      if feature[\"type\"] == \"Signal\":\n","        pos_cleavage_site = feature[\"location\"][\"end\"][\"value\"] #position of signal peptide cleavage site\n","    return [(protein_id, organism_name, kingdom, protein_length, pos_cleavage_site),fasta_line]"],"metadata":{"id":"HYtf2gzjC37Z","executionInfo":{"status":"ok","timestamp":1758286653119,"user_tz":-120,"elapsed":34,"user":{"displayName":"Aurora Mazzoni","userId":"09614232835557211324"}}},"execution_count":18,"outputs":[]},{"cell_type":"markdown","source":["#### Definition of a function to execute the positive dataset retrivial.\n","##### It takes as input the url and the 2 output file names"],"metadata":{"id":"vKEP-8yYBPYM"}},{"cell_type":"code","source":["def get_dataset_positive(search_url, output_file_tsv, output_file_fasta):\n","    filtered_json = []\n","    n_total, n_filtered = 0, 0\n","    # Run the API call in batches\n","    for batch, total in get_batch(search_url):\n","        # parse the JSON body of the response\n","        batch_json = json.loads(batch.text)\n","        # filter the entries\n","        for entry in batch_json[\"results\"]:\n","            n_total += 1\n","            # Check if the entry passes the filter\n","            if filter_entry_positive(entry):\n","                n_filtered += 1\n","                filtered_json.append(entry)\n","    print(n_total, n_filtered)\n","    with open(output_file_tsv, \"w\") as f1, open(output_file_fasta, \"w\") as f2:\n","      print(\"protein_id\", \"organism_name\", \"kingdom\", \"protein_length\", \"pos_cleavage_site\",sep=\"\\t\", file=f1)\n","      for entry in filtered_json:\n","          # Extract the fields of interest\n","          fields = extract_fields_positive(entry)[0]\n","          #Extract protein sequence\n","          seq = extract_fields_positive(entry)[1]\n","          # Print the fields in TSV format\n","          print(*fields, sep=\"\\t\", file=f1)\n","          # Print the sequence in FASTA format\n","          print(seq, file=f2)\n","      f1.close\n","      f2.close\n","    return"],"metadata":{"id":"SA3VmKetYKQh","executionInfo":{"status":"ok","timestamp":1758286653133,"user_tz":-120,"elapsed":5,"user":{"displayName":"Aurora Mazzoni","userId":"09614232835557211324"}}},"execution_count":19,"outputs":[]},{"cell_type":"markdown","source":["#### Output"],"metadata":{"id":"a2cDk_YYwYR3"}},{"cell_type":"code","source":["# We set the name of the output TSV file\n","output_positive_file_tsv = \"sp_positive.tsv\"\n","# And a name for the FASTA file\n","output_positive_file_fasta = \"sp_positive.fasta\"\n","\n","#Call the function to execute the whole script for positive dataset retrivial\n","get_dataset_positive(url_positive, output_positive_file_tsv, output_positive_file_fasta)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"SlTkNelqYWGE","executionInfo":{"status":"ok","timestamp":1758286661434,"user_tz":-120,"elapsed":8296,"user":{"displayName":"Aurora Mazzoni","userId":"09614232835557211324"}},"outputId":"9cae7d76-1385-4f3f-b2d2-24589e026bab"},"execution_count":20,"outputs":[{"output_type":"stream","name":"stdout","text":["2949 2932\n"]}]},{"cell_type":"markdown","source":["# ******negative dataset******"],"metadata":{"id":"kmw7oF4awfc0"}},{"cell_type":"code","source":["def get_dataset_negative(search_url, output_file_tsv, output_file_fasta):\n","    filtered_json = []\n","    n_total, n_filtered = 0, 0\n","    # Run the API call in batches\n","    for batch, total in get_batch(search_url):\n","        # parse the JSON body of the response\n","        batch_json = json.loads(batch.text)\n","        # filter the entries\n","        for entry in batch_json[\"results\"]:\n","            n_total += 1\n","            # Check if the entry passes the filter\n","            n_filtered += 1\n","            filtered_json.append(entry)\n","    print(n_total)\n","    with open(output_file_tsv, \"w\") as f1, open(output_file_fasta, \"w\") as f2:\n","      print(\"protein_id\", \"organism_name\", \"kingdom\", \"protein_length\", \"pos_cleavage_site\",sep=\"\\t\", file=f1)\n","      for entry in filtered_json:\n","          # Extract the fields of interest\n","          fields = extract_fields_negative(entry)[0]\n","          #Extract protein sequence\n","          seq = extract_fields_negative(entry)[1]\n","          # Print the fields in TSV format\n","          print(*fields, sep=\"\\t\", file=f1)\n","          # Print the sequence in FASTA format\n","          print(seq, file=f2)\n","      f1.close\n","      f2.close\n","    return\n","\n","\n","def extract_fields_negative(entry):\n","  # We extract Uniprot accession, organism name, eukaryotic kingdom, protein length, position of signal peptide cleavage site\n","  # We iterate over the features of the entry\n","    protein_id = entry[\"primaryAccession\"] #Uniprot accession\n","    organism_name = entry[\"organism\"][\"scientificName\"] #organism name\n","    kingdom = entry[\"organism\"][\"lineage\"][1] #eukaryotic kingdom\n","    protein_length = entry[\"sequence\"][\"length\"] #protein length\n","    helix_presence = \"5\" #presence of transmember helix in the first 90 residues\n","    helix_presence = False\n","    fasta_line = \">\"\n","    fasta_line += entry[\"primaryAccession\"]\n","    fasta_line += \"\\n\"\n","    fasta_line += entry[\"sequence\"][\"value\"] #protein sequence\n","    for feature in entry[\"features\"]:\n","      if feature[\"type\"] == \"Transmembrane\":\n","        helix_presence = True #presence of transmember helix in the first 90 residues\n","    return [(protein_id, organism_name, kingdom, protein_length, helix_presence),fasta_line]\n"],"metadata":{"id":"Cza4L7YCxFts","executionInfo":{"status":"ok","timestamp":1758286661447,"user_tz":-120,"elapsed":5,"user":{"displayName":"Aurora Mazzoni","userId":"09614232835557211324"}}},"execution_count":21,"outputs":[]},{"cell_type":"code","source":["# We set the name of the output TSV file\n","output_negative_file_tsv = \"sp_negative.tsv\"\n","# And a name for the FASTA file\n","output_negative_file_fasta = \"sp_negative.fasta\"\n","\n","url_negative = \"https://rest.uniprot.org/uniprotkb/search?format=json&query=%28fragment%3Afalse%29+AND+%28taxonomy_id%3A2759%29+AND+%28length%3A%5B40+TO+*%5D%29+NOT+%28ft_signal%3A*%29+AND+%28reviewed%3Atrue%29+AND+%28existence%3A1%29+AND+%28%28cc_scl_term_exp%3ASL-0091%29+OR+%28cc_scl_term_exp%3ASL-0191%29+OR+%28cc_scl_term_exp%3ASL-0173%29+OR+%28cc_scl_term_exp%3ASL-0209%29+OR+%28cc_scl_term_exp%3ASL-0204%29+OR+%28cc_scl_term_exp%3ASL-0039%29%29&size=500\"\n","\n","get_dataset_negative(url_negative, output_negative_file_tsv, output_negative_file_fasta)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"-UBAo5Fcwf0J","executionInfo":{"status":"ok","timestamp":1758286708276,"user_tz":-120,"elapsed":46824,"user":{"displayName":"Aurora Mazzoni","userId":"09614232835557211324"}},"outputId":"ef8968de-a685-4a1e-8341-415b41f91c79"},"execution_count":22,"outputs":[{"output_type":"stream","name":"stdout","text":["20615\n"]}]}]}